pkg/provisioning/provision_test.go

```diff
@@ -0,0 +1,511 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package provisioning
+
+import (
+	"context"
+	"testing"
+
+	"github.com/conduitio/conduit/pkg/connector"
+	connmock "github.com/conduitio/conduit/pkg/connector/mock"
+	"github.com/conduitio/conduit/pkg/foundation/cerrors"
+	"github.com/conduitio/conduit/pkg/foundation/database/inmemory"
+	"github.com/conduitio/conduit/pkg/foundation/log"
+	"github.com/conduitio/conduit/pkg/pipeline"
+	"github.com/conduitio/conduit/pkg/processor"
+	"github.com/conduitio/conduit/pkg/provisioning/mock"
+	"github.com/golang/mock/gomock"
+	"github.com/matryer/is"
+)
+
+var pipeline1 = PipelineConfig{
+	Status:      "running",
+	Name:        "pipeline1",
+	Description: "desc1",
+	Connectors: map[string]ConnectorConfig{
+		"con1": {
+			Type:   "source",
+			Plugin: "builtin:file",
+			Name:   "source",
+			Settings: map[string]string{
+				"path": "my/path/file1.txt",
+			},
+		},
+		"con2": {
+			Type:   "destination",
+			Plugin: "builtin:file",
+			Name:   "dest",
+			Settings: map[string]string{
+				"path": "my/path/file2.txt",
+			},
+			Processors: map[string]ProcessorConfig{
+				"proc1con": {
+					Type: "js",
+					Settings: map[string]string{
+						"additionalProp1": "string",
+					},
+				},
+			},
+		},
+	},
+	Processors: map[string]ProcessorConfig{
+		"proc1": {
+			Type: "js",
+			Settings: map[string]string{
+				"additionalProp1": "string",
+			},
+		},
+	},
+}
+
+var pipeline2 = PipelineConfig{
+	Status:      "running",
+	Name:        "pipeline2",
+	Description: "desc2",
+	Connectors: map[string]ConnectorConfig{
+		"con1": {
+			Type:   "source",
+			Plugin: "builtin:file",
+			Name:   "source",
+			Settings: map[string]string{
+				"path": "my/path/file1.txt",
+			},
+		},
+		"con2": {
+			Type:   "destination",
+			Plugin: "builtin:file",
+			Name:   "dest",
+			Settings: map[string]string{
+				"path": "my/path/file2.txt",
+			},
+		},
+	},
+}
+
+var pipeline3 = PipelineConfig{
+	Status:      "running",
+	Name:        "pipeline3",
+	Description: "desc3",
+	Connectors: map[string]ConnectorConfig{
+		"con1": {
+			Type:   "source",
+			Plugin: "builtin:file",
+			Name:   "source",
+			Settings: map[string]string{
+				"path": "my/path/file1.txt",
+			},
+		},
+		"con2": {
+			Type:   "destination",
+			Plugin: "builtin:file",
+			Name:   "dest",
+			Settings: map[string]string{
+				"path": "my/path/file2.txt",
+			},
+		},
+	},
+}
+
+func TestProvision_PipelineWithConnectorsAndProcessors(t *testing.T) {
+	is := is.New(t)
+	logger := log.Nop()
+	db := &inmemory.DB{}
+	_, ctx, _ := db.NewTransaction(context.Background(), true)
```

Comment:
It took me some time to figure out why we have this line - I assume we do this to get a context that will fit the expectation in the mock. It's a bit hacky and will break if we wrap the context and add something else to it other than the transaction, I'd rather just check that the context is not nil (e.g. `myMock.EXPECT().Call( gomock.Not(gomock.Nil()), ... )`). The best solution would be to have a matcher that makes sure the context contains a transaction, but that's a bit harder.